File: 2017/day04.py
#!/usr/bin/env python

def is_valid_one(line):
    """
    >>> is_valid_one("aa bb cc dd ee")
    True
    >>> is_valid_one("aa bb cc dd aa")
    False
    >>> is_valid_one("aa bb cc dd aaa")
    True
    """
    words = line.split()
    return max([words.count(x) for x in words]) == 1

def count_valid_passphrases(lines, is_valid):
    checks = [is_valid(x) for x in lines]
    return checks.count(True)

def sort_word(word):
    """
    >>> sort_word("asdfasdf")
    'aaddffss'
    """
    return "".join(sorted([x for x in word]))

def is_valid_two(line):
    """
    >>> is_valid_two("abcde fghij")
    True
    >>> is_valid_two("abcde xyz ecdab")
    False
    >>> is_valid_two("a ab abc abd abf abj")
    True
    >>> is_valid_two("iiii oiii ooii oooi oooo")
    True
    >>> is_valid_two("oiii ioii iioi iiio")
    False
    """
    words = line.split()
    words = [sort_word(x) for x in words]
    return max([words.count(x) for x in words]) == 1

def main():
    puzzle_input = [line.rstrip('\n') for line in open("day04_input.txt")]
    print "Part 1 Answer", count_valid_passphrases(puzzle_input, is_valid_one)
    print "Part 2 Answer", count_valid_passphrases(puzzle_input, is_valid_two)

if __name__ == '__main__':
    import doctest
    doctest.testmod()
    main()

File: 2017/adventofcode/lib/day04.ex
defmodule Day04 do
  import Adventofcode

  @moduledoc """
  Documentation for Day04.
  """

  @doc """
  iex> Day04.isvalid1("aa bb cc dd ee")
  true
  iex> Day04.isvalid1("aa bb cc dd aa")
  false
  iex> Day04.isvalid1("aa bb cc dd aaa")
  true
  """
  def isvalid1(passphrase) do
    pre = passphrase |> String.split()
    post = pre |> Enum.sort() |> Enum.dedup()
    length(pre) == length(post)
  end

  def part1(phrases) do
    phrases |> Enum.count(&isvalid1/1)
  end

  @doc """
  iex> Day04.isvalid2("abcde fghij")
  true
  iex> Day04.isvalid2("abcde xyz ecdab")
  false
  iex> Day04.isvalid2("a ab abc abd abf abj")
  true
  iex> Day04.isvalid2("iiii oiii ooii oooi oooo")
  true
  iex> Day04.isvalid2("oiii ioii iioi iiio")
  false
  """
  def isvalid2(passphrase) do
    pre =
      passphrase
      |> String.split()
      |> Enum.map(fn x -> String.graphemes(x) |> Enum.sort() |> Enum.join() end)

    post = pre |> Enum.sort() |> Enum.dedup()
    length(pre) == length(post)
  end

  def part2(phrases) do
    phrases |> Enum.count(&isvalid2/1)
  end

  def run do
    input = read_input(4)
    answer(1, 451, part1(input))
    answer(2, 223, part2(input))
  end
end
